flocker/node/agents/test/blockdevicefactory.py

```diff
@@ -94,28 +102,64 @@ def get_blockdeviceapi_args(provider):
     with open(config_file_path) as config_file:
         config = safe_load(config_file.read())
 
-    section = config[provider.name]
+    section = config.get(platform_name)
+    if section is None:
+        raise InvalidConfig(
+            "The requested cloud platform "
+            "was not found in the configuration file. "
+            "Platform: %s, "
+            "Configuration File: %s" % (platform_name, config_file_path)
+        )
+
+    provider_name = section.get('provider', platform_name)
+    try:
+        provider_environment = ProviderType.lookupByName(provider_name)
+    except ValueError:
+        raise InvalidConfig(
+            "Unsupported provider. "
+            "Supplied provider: %s, "
+            "Available providers: %s" % (
+                provider_name,
+                ', '.join(p.name for p in ProviderType.iterconstants())
+            )
+        )
+
+    if provider_environment != provider:
+        raise InvalidConfig(
+            "The requested cloud provider (%s) is not the provider running "
+            "the tests (%s)." % (provider.name, provider_environment.name)
+        )
+
     cls, get_kwargs = _BLOCKDEVICE_TYPES[provider]
     kwargs = dict(cluster_id=uuid4())
     kwargs.update(get_kwargs(**section))
     return cls, kwargs
 
 
-def _rackspace_session(username, key, **kwargs):
-    """
-    Create a Keystone session capable of authenticating with Rackspace.
+from keystoneclient.auth import get_plugin_class
+# from keystoneclient.exceptions import NoMatchingPlugin
 
-    :param unicode keyname: A RackSpace API username.
-    :param unicode key: A RackSpace API key.
 
-    :return: A ``keystoneclient.session.Session``.
-    """
-    auth = RackspaceAuth(
-        auth_url=RACKSPACE_AUTH_URL,
-        username=username,
-        api_key=key
-    )
-    return Session(auth=auth)
+def _openstack_auth_from_config(**config):
```

Comment:
This might be premature generalization.  I don't know that we've seen a second openstack cloud where we can use sessions for authentication yet.